atmat/atphysics/Radiation/FDW.c

```diff
@@ -0,0 +1,572 @@
+	   BDIFF - cumulative Ohmi's diffusion is initialized to 0
+	   BDIFF is preallocated 1 dimensional array to store 6-by-6 matrix 
+	   columnwise.
+	*/
+	
+	int m;	
+	double  *MKICK, *BKICK;
+
+	/* 4-th order symplectic integrator constants */
+	double dl1,dl0;
+    int Niter = Nstep*(le/Lw);
+    double SL = Lw/Nstep;
+	double zEndPointH[2];
+    double zEndPointV[2];
+	double ax,ay,axpy,aypx;
+	double B[2];
+	double E0G;
+    struct gwigR pWig;
+    int flag;       
+    
+	dl1 = x1*SL;
+    dl0 = x0*SL;
+	
+	/* Allocate memory for thin kick matrix MKICK
+	   and a diffusion matrix BKICK
+	*/
+ 	MKICK = (double*)mxCalloc(36,sizeof(double));
+	BKICK = (double*)mxCalloc(36,sizeof(double));
+	for(m=0; m < 6; m++)
+		{	MKICK[m] = 0;
+			BKICK[m] = 0;
+		}
+	
+	/* Transform orbit to a local coordinate system of an element
+       BDIFF stays zero	*/
+    if(pt1)
+        ATaddvv(orbit_in,pt1);	
+    if(PR1)
+        ATmultmv(orbit_in,PR1);	
+
+   
+    /*Generate the wiggler*/ 
+
+  
+  zEndPointH[0] = 0;
+  zEndPointH[1] = le;
+  zEndPointV[0] = 0;
+  zEndPointV[1] = le;
+  
+  E0G = E0 / 1e9;
+
+		  
+  GWigInit2(&pWig, E0G, le, Lw, Bmax, Nstep, Nmeth,NHharm,NVharm,0,0,zEndPointH,zEndPointV,pBy,pBx,pt1,pt2,PR1,PR2);
+    
+ /*   GWigInit(&pWig, le, Lw, Bmax, Nstep, Nmeth,NHharm,NVharm,pBy,pBx,
+              E0,pt1,pt2,PR1,PR2);
+    
+ */   
+	/* Propagate orbit_in and BDIFF through a 4-th orderintegrator */
+        
+    flag=1;
+    GWigGauge(&pWig, orbit_in, flag);
+ 
+    GWigAx(&pWig, orbit_in, &ax, &axpy);
+    GWigAy(&pWig, orbit_in, &ay, &aypx);
+    GWigB(&pWig, orbit_in, B);
+
+    orbit_in[1] -= ax;
+    orbit_in[3] -= ay;
+    GWigRadiationKicks(&pWig, orbit_in, B, SL);
+    orbit_in[1] += ax;
+    orbit_in[3] += ay;
+	
+    for (m = 1; m <= Niter; m++ )   /* Loop over slices	*/
+		{	   
+
+			   wigglerM(&pWig, orbit_in, dl1, MKICK);
+			   wigglerB(&pWig, orbit_in, dl1, E0, BKICK);
+			   ATsandwichmmt(MKICK,BKICK);
+			   ATaddmm(BKICK,BDIFF);
+			   GWigMap_2nd(&pWig, orbit_in, dl1);
+                
+			   wigglerM(&pWig, orbit_in, dl0, MKICK);
+			   wigglerB(&pWig, orbit_in, dl0, E0, BKICK);
+			   ATsandwichmmt(MKICK,BKICK);
+			   ATaddmm(BKICK,BDIFF);
+               GWigMap_2nd(&pWig, orbit_in, dl0);
+               
+			   wigglerM(&pWig, orbit_in, dl1, MKICK);
+			   wigglerB(&pWig, orbit_in, dl1, E0, BKICK);
+			   ATsandwichmmt(MKICK,BKICK);
+			   ATaddmm(BKICK,BDIFF);
+               GWigMap_2nd(&pWig, orbit_in, dl1);
+			   
+			   GWigAx(&pWig, orbit_in, &ax, &axpy);
+			   GWigAy(&pWig, orbit_in, &ay, &aypx);
+               GWigB(&pWig, orbit_in, B);
+			   orbit_in[1] -= ax;
+               orbit_in[3] -= ay;
+               GWigRadiationKicks(&pWig, orbit_in, B, SL);
+               orbit_in[1] += ax;
+               orbit_in[3] += ay;	
+		}      
+        
+        flag=-1;
+        GWigGauge(&pWig, orbit_in, flag);
+       
+		
+        if(PR2)
+            ATmultmv(orbit_in,PR2);	

```

Comment:
You should propagate the diffusion matrix through the rotation matrix. You are right that at input, propagating a zero matrix is useless, but I think here it should be done.

Reply:
Be careful, Yrot is a rotation around the Y axis. Here we are talking of a rotation around the Z axis. I think that the propagation consists just in applying the "ATsandwich" to the rotation matrix (R1 or R2).